[Mailer][MailJet] Fix forbidden headers case-sensitive comparison

benjamintoussaint opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #44000
| License       | MIT

We need to ensure that forbidden headers are not sent, regardless of their case, e.g. "Sender" and "sender" should both be forbidden.

benjamintoussaint committed:
[Mailer] Fix MailJet bridge forbidden headers case-sensitive comparison

We need to ensure that forbidden headers are not sent, regardless of their case, e.g. "Sender" and "sender" should both be forbidden.

benjamintoussaint commented:
¯\_(ツ)_/¯

stof commented on src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:
```diff
@@ -144,7 +147,8 @@ private function getPayload(Email $email, Envelope $envelope): array
             }
-            if (\in_array($header->getName(), self::FORBIDDEN_HEADERS, true)) {
+
+            if ($this->isHeaderForbidden($header)) {
```
A better solution is to use the key of the array returned by `$email->getHeaders()->all()`. In that array, keys are always lowercase, while `$header->getName()` returns the original name (which we use when sending them).

Note that the logic for `$convertConf` should also be updated to handle name case insensitively by using the array key as header name.

benjamintoussaint replied:
Thank you for your review; I've pushed the requested changes.

benjamintoussaint committed:
[Mailer] Use all headers key in loop as lowercase header name

This follows the PR review.

OskarStark commented on src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:
```diff
@@ -51,6 +52,8 @@ class MailjetApiTransport extends AbstractApiTransport
         'X-Mailjet-TrackOpen' => ['TrackOpens', 'string'],
     ];
 
+    private ?array $forbiddenHeadersLowercase = null;
```
do we really need this property?

benjamintoussaint replied:
No, I just didn't want to apply `strtolower()` at each loop round.

benjamintoussaint committed:
[Mailer] Use a local variable to compare header names

This follows the PR review.

stof commented on src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:
```diff
@@ -139,12 +139,15 @@ private function getPayload(Email $email, Envelope $envelope): array
 
-        foreach ($email->getHeaders()->all() as $header) {
-            if ($convertConf = self::HEADER_TO_MESSAGE[$header->getName()] ?? false) {
+        $forbiddenHeadersLowercase = \array_map('strtolower', self::FORBIDDEN_HEADERS);
```
this should be removed. Instead, the constant should be updated to contain lowercase names

benjamintoussaint replied:
I think that it's best to keep the original forbidden headers array as it is for readability, what do you think?

```
'Date', 'X-CSA-Complaints', 'Message-Id', 'X-MJ-StatisticsContactsListID'
```

VS 

```
'date', 'x-csa-complaints', 'message-id', 'x-mj-statisticscontactslistid',
```

stof replied:
Storing the constant in a format that cannot be used directly does not make sense to me.

Btw, storing it in lowercase is already what we do in other bridge packages

benjamintoussaint replied:
Alright then, I'll change that :-)

stof commented on src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:
```diff
@@ -139,12 +139,15 @@ private function getPayload(Email $email, Envelope $envelope): array
+        $forbiddenHeadersLowercase = \array_map('strtolower', self::FORBIDDEN_HEADERS);
+
+        foreach ($email->getHeaders()->all() as $headerName => $header) {
+            if ($convertConf = self::HEADER_TO_MESSAGE[$headerName] ?? false) {
```
the constant must be updated to use lowercased keys. Otherwise, this is broken as the keys won't be found in the array.

stof replied:
btw, this is the reason why tests are failing.

benjamintoussaint replied:
Don't you think that we need both uppercase and lowercase constants instead?
This risk by defining them all to lowercase is to have the very same issue with an uppercase header.

stof replied:
`$headerName` is always a lowercase string here (that's precisely why I asked you to use the array key instead of `$header->getName()`.

See the tests that use the title-case names, which are now broken by your change, but will get fixed if you update the constant.

benjamintoussaint replied:
Thanks for pointing that out, I have pushed the needed changes.

benjamintoussaint committed:
[Mailer] Fix header to message keys case discrepancy

This follows the PR review.

benjamintoussaint committed:
[Mailer] Fix header to message keys case discrepancy for keys

It's preferable to keep the constant with 'TitleCase' for readability, so we make all array keys lowercase prior to looping through the headers.

This follows the PR review.

benjamintoussaint committed:
Fix code style following PR check

benjamintoussaint committed:
[Mailer] Simplify case conversion logic to its base

This follows the PR review.

fabpot commented:
Thank you @benjamintoussaint.
